src/normalization.md

```diff
@@ -0,0 +1,314 @@
+}
+```
+
+Ultimately this means that we have no guarantee that aliases in the type system terminate. As aliases may only diverge for some specific arguments to generic parameters, it also means that only fully concrete aliases can be checked to terminate, which means that codegen/const evaluation is the only time that aliases can reliably be checked to terminate:
+```rust
+trait Trait {
+    type Diverges<U: Trait>;
+}
+impl<T: ?Sized> Trait for T {
+    type Diverges<U: Trait> = <U as Trait>::Diverges<U>;
+}
+
+fn foo<T: Trait>() {
+    let a: T::Diverges<T>;
+}
+
+fn main() {
+    foo::<()>();
+}
+```
+In this example we only encounter an error from the diverging alias during codegen of `foo::<()>`, if the call to `foo` is removed then no compilation error will be emitted.
+
+### Opaque Types
+
+Opaque types are a relatively special kind of alias, and are covered in their own chapter: [Opaque types](../opaque-types-type-alias-impl-trait.md).
+
+### Const Aliases
+
+Unlike type aliases, const aliases are not represented directly in the type system, instead const aliases are always an anonymous body containing a path expression to a const item. This means that the only "const alias" in the type system is an anonymous unevaluated const body.
+
+As such there is no `ConstKind::Alias(AliasCtKind::Projection/Inherent/Free, _)`, instead we only have `ConstKind::Unevaluated` which is used for representing anonymous constants.
+
+```rust
+fn foo<const N: usize>() {}
+
+const FREE_CONST: usize = 1 + 1;
+
+fn bar() {
+    foo::<{ FREE_CONST }>();
+    // The const arg is represented with some anonymous constant:
+    // ```pseudo-rust
+    // const ANON: usize = FREE_CONST; 
+    // foo::<ConstKind::Unevaluated(DefId(ANON), [])>();
+    // ```
+}
+```
+
+This is likely to change as const generics functionality is improved, for example `feature(associated_const_equality)` and `feature(min_generic_const_args)` both require handling const aliases similarly to types (without an anonymous constant wrapping all const args).
+
+## Normalization
+
+There are two main "parts" to how normalization exists in the compiler:
+- The core logic responsible for turning an alias into its underlying type
+- The various entry points to the core normalization logic that expose real APIs
+
+### Core normalization logic
+
+Normalizing aliases is a little bit more nuanced than replacing the alias with whatever it was defines as being equal to in its definition; the result of normalizing an alias should either be a rigid type or an inference variable (which will later be inferred to a rigid type). To accomplish this we do two things:
+
+First, when normalizing an ambiguous alias it is normalized to an inference variable instead of leaving it as-is, this has two main effects: 
+- Even though an inference variable is not a rigid type, it will always wind up inferred *to* a rigid type so we ensure that the result of normalization will not need to be normalized again
```

Comment:
shallow vs deep normalization. idk how to best introduce this, but it only guarantees that the "outer type constructor" is rigid.

you could normalize to `fn(<&() as Trait>::Assoc>)` after all which contains a normalizeable alias.

I think before explaining *how* normalization works, it may make sense to introduce "structural normalization" and "deep normalization"

Reply:
Yeah it feels like its missing a section specifically calling out shallow vs deep norm :thinking: It's awkward because the old solver kinda likes to pretend that normalization is deep, yet `deeply_normalize` exists and is arguably just as "deep" as shallow norm in the old solver.

Even in the new solver "shallow" normalization is largely deep norm with the exception of aliases in binders </3

Reply:
Or well, deeply_normalize *is* arguably more deep in the sense that you never have to worry about encountering non-rigid aliases inside of the term anymore. But that feels so subtle given it only *really* matters in the edge case of hr aliases :sweat_smile: